src/streamer.rs

```diff
@@ -0,0 +1,338 @@
+use std::sync::{Arc, Mutex, RwLock};
+use std::sync::mpsc::{Receiver, Sender};
+use std::time::Duration;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, UdpSocket};
+use std::thread::{spawn, JoinHandle};
+use result::{Error, Result};
+
+const BLOCK_SIZE: usize = 1024 * 8;
+const PACKET_SIZE: usize = 256;
+
+#[derive(Clone)]
+pub struct Packet {
+    pub data: [u8; PACKET_SIZE],
+    pub size: usize,
+    pub addr: [u16; 8],
+    pub port: u16,
+    pub v6: bool,
+}
+
+impl Default for Packet {
+    fn default() -> Packet {
+        Packet {
+            data: [0u8; PACKET_SIZE],
+            size: 0,
+            addr: [0u16; 8],
+            port: 0,
+            v6: false,
+        }
+    }
+}
+impl Packet {
+    pub fn get_addr(&self) -> SocketAddr {
+        if !self.v6 {
+            let ipv4 = Ipv4Addr::new(
+                self.addr[0] as u8,
+                self.addr[1] as u8,
+                self.addr[2] as u8,
+                self.addr[3] as u8,
+            );
+            SocketAddr::new(IpAddr::V4(ipv4), self.port)
+        } else {
+            let ipv6 = Ipv6Addr::new(
+                self.addr[0],
+                self.addr[1],
+                self.addr[2],
+                self.addr[3],
+                self.addr[4],
+                self.addr[5],
+                self.addr[6],
+                self.addr[7],
+            );
+            SocketAddr::new(IpAddr::V6(ipv6), self.port)
+        }
+    }
+
+    pub fn set_addr(&mut self, a: &SocketAddr) {
+        match a {
+            &SocketAddr::V4(v4) => {
+                let ip = v4.ip().octets();
+                self.addr[0] = ip[0] as u16;
+                self.addr[1] = ip[1] as u16;
+                self.addr[2] = ip[2] as u16;
+                self.addr[3] = ip[3] as u16;
+                self.port = a.port();
+            }
+            &SocketAddr::V6(v6) => {
+                self.addr = v6.ip().segments();
+                self.port = a.port();
+                self.v6 = true;
+            }
+        }
+    }
+}
+
+pub struct PacketData {
+    pub packets: Vec<Packet>,
+}
+
+type SharedPacketData = Arc<RwLock<PacketData>>;
+type Recycler = Arc<Mutex<Vec<SharedPacketData>>>;
+
+impl PacketData {
+    pub fn new() -> PacketData {
+        PacketData {
+            packets: vec![Packet::default(); BLOCK_SIZE],
+        }
+    }
+    fn run_read_from(&mut self, socket: &UdpSocket) -> Result<usize> {
+        self.packets.resize(BLOCK_SIZE, Packet::default());
+        let mut i = 0;
+        for p in self.packets.iter_mut() {
```

Comment:
enumerate()?

Reply:
This is really dumb, but, if i used an enumerate, i would still need a mut to return the value outside of the loop.  If i had a return instead of a break, then the return statement at the end of the function would be unreachable.